Ignore list

kandski opened the pull request.

kandski committed:
add ignore list prototype

kandski committed:
Fix code to comply to PEP8

kandski committed:
Fix unused imports

matejak commented on src/iwant_bot/ignore.py:
```diff
@@ -0,0 +1,23 @@
+    def __init__(self):
+        self.ignored_users = collections.defaultdict(list)
+
+    def ignore(self, user, ignored):
```
I have a naming brainstorming result: As the class name is `IgnoreList`, adding a "X wants to ignore Y" relation looks like this:
```
ignore_list = IgnoreList()
ignore_list.ignore(X, Y)
```
It could look like
```
ignore_list.add(X, Y)
ignore_list.apply(X, IGNORES, Y)
```
and not ignoring any more
```
ignore_list.remove(X, Y)
ignore_list.apply(X, DOESNT_IGNORE, Y)
```
Like it or forget it.

kandski replied:
Good idea.

matejak commented on src/iwant_bot/ignore.py:
```diff
@@ -0,0 +1,23 @@
+    def who(self, user):
+        return self.ignored_users[user]
+
+    def is_ignored(self, user, ignored):
```
Probably OK, but what about `user_ignores(self, user, potentially_ignored)`?

kandski replied:
A bit redundant in this case for me. Imho "ignored" is well understandable in this case.

kandski replied:
But I agree on that user_ignores part 👌

matejak commented on src/iwant_bot/pool.py:
```diff
@@ -20,6 +21,7 @@ def __init__(self, storage):
         self._time_relevant_requests = set()
         self._blacklisted_requests = set()
         self._time_conflicting_requests = set()
+        self.users = IgnoreList()
```
I would not guess that the attribute `users` has an `IgnoreList` instance assigned to it. Why not to call it `ignore_list`?

kandski replied:
It will be better, naming in hurry ™

matejak commented on src/iwant_bot/ignore.py:
```diff
@@ -0,0 +1,23 @@
+
+class IgnoreList(object):
+    def __init__(self):
+        self.ignored_users = collections.defaultdict(list)
```
IgnoreList is a nicely designed class, there is no real need to expose `ignored_users` as a public attribute.

kandski replied:
You are right.

matejak commented on src/iwant_bot/pool.py:
```diff
@@ -43,13 +45,20 @@ def update_requests_from_storage(self):
-            destination.append(paired)
+        if len(source) > 1:
+            for request1 in source:
+                source.remove(request1)
```
I don't have a good experience with iterating over a container while removing its elements: https://stackoverflow.com/questions/10665591/how-to-remove-list-elements-in-a-for-loop-in-python

kandski replied:
I will take a look at it. Soon. I'm thankful for your review. It was a bit unhappy and first working solution, but it's not perfect.

matejak commented on tests/unit/test_pool.py:
```diff
@@ -21,20 +22,37 @@ def make_filled_request_pool():
+    pool.users.ignore('christopher', 'andrew')
+    pool.users.ignore('olivia', 'andrew')
+    pool.users.ignore('olivia', 'martin')
+    for activity in pool.activity_list:
```
Nice ignore list, but the pool needs to be expanded. As there are 2 if statements in the ignorelist-related `make_pairs` method, at least 4 checks of the method are needed to test it fully.

kandski replied:
Good point.

matejak replied:
I will try to get the coverage report working over the weekend. That will help a lot to write tests.

kandski committed:
update logic of ignore list structure

kandski committed:
update tests of ignore list and related pool

petrjoachim commented on src/iwant_bot/pool.py:
```diff
@@ -43,33 +47,26 @@ def update_requests_from_storage(self):
 
     def make_pairs(self, activity):
+        ignore_list = self.ignore_list
         source = self.req_by_activities[activity]
```
`source` => `requests`

petrjoachim commented on src/iwant_bot/pool.py:
```diff
@@ -43,33 +47,26 @@ def update_requests_from_storage(self):
+                    source.remove(request2)
+                    destination.append(paired)
+                    break
+                source.append(request1)
```
Transform this if/for/for piece to linear code using `zip` `filter` and make uniqueness by overriding `__eq__`
```
requests = set(requests)
pairs = map(make_pair, zip(requests))
pairs = filter(exclude_self, pairs)
pairs = filter(exclude_ignored, pairs)
...
remaining_requests = requests - flatten(pairs)
```

petrjoachim reviewed (changes requested):
There are some way to make this more pythonic

kandski committed:
change the way of creating pairs

change the way of creating pairs and make it more pythonic
add possibility to create a groups
add group ignore check

kandski committed:
add end to the new line of ignore.py

kandski commented:
Closes #46
